corehq/apps/es/README.rst

```diff
@@ -0,0 +1,220 @@
+  uses, which guarantees that any possible problems encountered by a third party
+  hoster are not outside the Dimagi main track.
+
+
+.. TODO: Future adapter documentation
+..
+.. - adapter patterns
+.. - querying with adapters
+.. - writing tests (use ``es_test``)
+
+----
+
+Elastic Client Adapters
+-----------------------
+
+The ``corehq.apps.es.client`` module encapsulates the CommCare HQ Elasticsearch
+client adapters. It implements a high-level Elasticsearch client protocol
+necessary to accomplish all interactions with the backend Elasticsearch cluster.
+Client adapters are split into two usage patterns, the "Management Adapter" and
+"Document Adapters".
+
+.. toctree::
+
+    Management Adapter
+    Document Adapters
+    Code Documentation
+
+
+Management Adapter
+''''''''''''''''''
+
+There is only one management adapter, ``ElasticManageAdapter``. This adapter is
+used for performing all cluster management tasks such as creating and updating
+indices and their mappings, changing index settings, changing cluster settings,
+etc.  This functionality is split into a separate class for a few reasons:
+
+1. The management adapter is responsible for low-level Elastic operations which
+   document adapters should never be performing because the scope of a document
+   adapter does not extend beyond a single index.
+2. Elasticsearch 5+ implements security features which limit the kinds of
+   operations a connection can be used for. The separation in these client
+   adapter classes is designed to fit into that model.
+
+The management adapter does not need any special parameters to work with, and
+can be instantiated and used directly:
+
+.. code-block:: python
+
+    adapter = ElasticManageAdapter()
+    adapter.index_create("books")
+    mapping = {"properties": {
+        "author": {"type": "text"},
+        "title": {"type": "text"},
+        "published": {"type": "date"},
+    }}
+    adapter.index_put_mapping("books", "book", mapping)
+    adapter.index_refresh("books")
+    adapter.index_delete("books")
+
+
+Document Adapters
+'''''''''''''''''
+
+Document adapters are created on a per-index basis and include specific
+properties and functionality necessary for maintaining a single type of "model"
+document in a single index.  Each index in Elasticsearch needs to have a
+cooresponding ``ElasticDocumentAdapter`` subclass which defines how the Python
+model is applied to that specific index.  At the very least, a document adapter
+must define the following:
+
+- An ``_index_name`` attribute whose value is the name of the Elastic index
+  used by the adapter. This attribute must be private to support proper index
+  naming between production code and tests.
+- A ``type`` attribute whose value is the name is the Elastic ``_type`` for
+  documents used by the adapter.
+- A ``mapping`` which defines the structure and properties for documents managed
+  by the adapter.
```

Comment:
What's the purpose of the mapping in the document adapter?

Reply:
Sure, but the document adapters don't need to know about the mapping unless you plan to add query validation or something like that right?

Reply:
Not in this PR. The multiplexing implementation will need it on the adapter to support things like mapping changes.

Reply:
I'd like more info on this piece too please :)